src/components/CodeEditor/index.jsx

```diff
@@ -57,29 +54,34 @@ export const CodeEditor = ({ code, children, canvasId }) => {
         return null;
       },
     });
-
+    const parsedCode = functionToCodeString(codeProp);
     const editor = (editorRef.current = new EditorView({
-      doc: functionToCodeString(code),
+      doc: parsedCode,
       extensions: [
         basicSetup,
         javascript(),
         javascriptLanguage.data.of({
           autocomplete: scopeCompletionSource(globalThis),
         }),
-        listenChangesExtension,
+        onChangeHook,
       ],
       parent: divRef.current,
     }));
-
+    runCallback([parsedCode]);
     return () => editor.destroy();
-  }, [runCallback]);
+  }, []);
 
   useEffect(() => {
-    runCallback();
-  }, []);
+    const handler = () => console.log('TODO: handle disposing gracefully');
+    window.addEventListener('canvas_dispose', handler);
+    return () => window.removeEventListener('canvas_dispose', handler);
+  });
 
   return (
     <>
+      <Helmet>
+        <script type="module">{code}</script>
```

Comment:
instead of eval

Reply:
not sure if there is a difference.
Is probably just slower?

Doesn't remove the problem of evaluating code written from the user.
Wraps the evaluation in 2-10 function wrappers.